cmd/bnsd/x/username/handlers.go

```diff
@@ -0,0 +1,165 @@
+		Owner:    owner,
+	}
+	if _, err := h.bucket.Put(db, msg.Username.Bytes(), &token); err != nil {
+		return nil, errors.Wrap(err, "cannot store token")
+	}
+	return &weave.DeliverResult{Data: msg.Username.Bytes()}, nil
+}
+
+func (h *registerUsernameTokenHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*RegisterUsernameTokenMsg, error) {
+	var msg RegisterUsernameTokenMsg
+	if err := weave.LoadMsg(tx, &msg); err != nil {
+		return nil, errors.Wrap(err, "load msg")
+	}
+
+	switch err := h.bucket.Has(db, msg.Username.Bytes()); {
+	case err == nil:
+		return nil, errors.Wrapf(errors.ErrDuplicate, "username %q already registered", msg.Username)
+	case errors.ErrNotFound.Is(err):
+		// All good. Username is not taken yet.
+	default:
+		return nil, errors.Wrap(err, "cannot check if username is unique")
+	}
+	return &msg, nil
+}
+
+type transferUsernameTokenHandler struct {
+	auth   x.Authenticator
+	bucket orm.ModelBucket
+}
+
+func (h *transferUsernameTokenHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	if _, _, err := h.validate(ctx, db, tx); err != nil {
+		return nil, err
+	}
+	return &weave.CheckResult{GasAllocated: transferUsernameTokenCost}, nil
+}
+
+func (h *transferUsernameTokenHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, token, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	token.Owner = msg.NewOwner
+	if _, err := h.bucket.Put(db, msg.Username.Bytes(), token); err != nil {
+		return nil, errors.Wrap(err, "cannot store token")
+	}
+	return &weave.DeliverResult{Data: msg.Username.Bytes()}, nil
+}
+
+func (h *transferUsernameTokenHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*TransferUsernameTokenMsg, *UsernameToken, error) {
+	var msg TransferUsernameTokenMsg
+	if err := weave.LoadMsg(tx, &msg); err != nil {
+		return nil, nil, errors.Wrap(err, "load msg")
+	}
+
+	var token UsernameToken
+	if err := h.bucket.One(db, msg.Username.Bytes(), &token); err != nil {
+		return nil, nil, errors.Wrap(err, "cannot get token from database")
+	}
+
+	if !h.auth.HasAddress(ctx, token.Owner) {
+		return nil, nil, errors.Wrap(errors.ErrUnauthorized, "only the token owner can execute this operation")
+	}
+
+	return &msg, &token, nil
+}
+
+type changeUsernameTokenTargetsHandler struct {
+	auth   x.Authenticator
+	bucket orm.ModelBucket
+}
+
+func (h *changeUsernameTokenTargetsHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	if _, _, err := h.validate(ctx, db, tx); err != nil {
+		return nil, err
+	}
+	return &weave.CheckResult{GasAllocated: changeUsernameTokenTargetCost}, nil
+}
+
+func (h *changeUsernameTokenTargetsHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, token, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	token.Targets = msg.NewTargets
```

Comment:
I remember us talking about not moving complexity to the frontend and using a Patch here. I haven't really followed the conversation after that, but I assume @alpe agreed with this approach or otherwise this change here is missing.

Reply:
This is discussed in https://github.com/iov-one/weave/pull/771#discussion_r294515327

Reply:
Yeah, so there was no further discussion. 

I'll be honest and say that to me the current implementation works just fine. The PATCH method does not really resolve issues with stuff being executed sequentially, because a PATCH can possibly contain all the same changes as POST and lead to similar issues, because the client will be using an outdated version of information stored in buckets.

So to me it boils down to complexity on the client vs server. And as I look at how this should properly be implemented on the client (and that's introspection of the current values to decide what the patch would be) - it looks like we'll have a similar amount of code either way.
Which leads me to conclusion that simplified logic on the backend won't hurt.